Sources/TensorFlow/Operators/LinearAlgebra.swift

```diff
@@ -266,3 +266,122 @@ public extension Tensor where Scalar: TensorFlowFloatingPoint {
+    if matrix.rank < rhs.rank {
+        let leadingDims = extractLeadingDims(rhs.shape, matrix.shape, ignoreLast: 2)
+        let matchedMatrix = matrix.broadcasted(to: leadingDims + matrix.shape)
+        return _Raw.matrixTriangularSolve(matrix: matchedMatrix, rhs: rhs, lower: lower, adjoint: adjoint)
+    } else if matrix.rank > rhs.rank {
+        let leadingDims = extractLeadingDims(rhs.shape, matrix.shape, ignoreLast: 2)
+        let matchedRhs = rhs.broadcasted(to: leadingDims + rhs.shape)
+        return _Raw.matrixTriangularSolve(matrix: matrix, rhs: matchedRhs, lower: lower, adjoint: adjoint)
+    }
+    return _Raw.matrixTriangularSolve(matrix: matrix, rhs: rhs, lower: lower, adjoint: adjoint)
+}
+
+@inlinable
+@derivative(of: triangularSolve)
+internal func _vjpTriangularSolve<T: TensorFlowFloatingPoint>(
+    matrix: Tensor<T>,
+    rhs: Tensor<T>,
+    lower: Bool = true,
+    adjoint: Bool = false
+) -> (value: Tensor<T>, pullback: (Tensor<T>) -> (Tensor<T>, Tensor<T>)) {
+    let broadcastMatrix: Bool = matrix.rank < rhs.rank
+    let broadcastRhs: Bool = matrix.rank > rhs.rank
+
+    var matchedMatrix: Tensor<T> = matrix
+    if broadcastMatrix {
+        let leadingDims = extractLeadingDims(rhs.shape, matrix.shape, ignoreLast: 2)
+        matchedMatrix = matrix.broadcasted(to: leadingDims + matrix.shape)
+    }
+    
+    let x = triangularSolve(matrix: matchedMatrix, rhs: rhs, lower: lower, adjoint: adjoint)
+    let triangularSolveGrad = { (gradient: Tensor<T>) -> (Tensor<T>, Tensor<T>) in
+        let rhsGrad = triangularSolve(matrix: matchedMatrix, rhs: gradient, lower: lower, adjoint: !adjoint)
+        let (left, right) = adjoint ? (x, rhsGrad) : (rhsGrad, x)
+        let matrixGrad = -matmul(left, transposed: false, right, transposed: true)
+        let triMatrixGrad = lower
+            ? matrixGrad.bandPart(subdiagonalCount: -1, superdiagonalCount: 0)
+            : matrixGrad.bandPart(subdiagonalCount: 0, superdiagonalCount: -1)
+        if broadcastMatrix {
+            return (triMatrixGrad.unbroadcasted(to: matrix.shape), rhsGrad)
+        } else if broadcastRhs {
+            return (triMatrixGrad, rhsGrad.unbroadcasted(to: rhs.shape))
+        }
+        return (triMatrixGrad, rhsGrad)
+    }
+    return (x, triangularSolveGrad)
+}
+
+
+/// Returns leading dimension from tensors shapes.
+/// Example:
+/// The leading dimension for tensors with shapes `[N, M, K]` and `[M, K]`
+/// is `[N]`. In cases, when tensors shapes do not have common parts from the end, e.g.
+/// `[N, M, K]` and `[N, M]`, it will raise an error.
+///
+/// - Parameters:
+///     - left: Left tensor.
+///     - right: Right tensor.
+///     - ignoreLast: Number of dimensions from the end to ignore. Default value is zero.
+/// - Precondition: Left tensor rank cannot be lower than `ignoreLast`.
+/// - Precondition: Right tensor rank cannot be lower than `ignoreLast`.
+@inlinable
+public func extractLeadingDims<T: TensorFlowNumeric>(
```

Comment:
Please make both `extractLeadingDims` on functions `internal`. We don't want to expose these functions to users - please keep that in mind when deciding whether to make any function `public`.

I know these functions are used in `LinearAlgebraTests.swift` as well, but that test has `@testable import TensorFlow`, which should allow it to access internal functions. That's the proper way to handle "functions that should be private but need to called from tests".

Reply:
The initial thinking was that these functions could be useful for users and other parts of the project. But, I don't mind to make them internal. If needed they can be exported after.

Reply:
> The initial thinking was that these functions could be useful for users and other parts of the project. But, I don't mind to make them internal. If needed they can be exported after.

It's not appropriate to expose these APIs with their current non-idiomatic naming. Public APIs should be orthogonal (and somewhat simple, we should try hard not to expose multiple APIs with overlapping functionality), named appropriately, and defined at an appropriate location.